Firestore/core/src/core/target.cc

```diff
@@ -41,6 +75,254 @@ bool Target::IsDocumentQuery() const {
+          field_value_map[segment.field_path().CanonicalString()] =
+              model::DeepClone(field_filter.value());
+          break;
+        case FieldFilter::Operator::NotIn:
+        case FieldFilter::Operator::NotEqual:
+          field_value_map[segment.field_path().CanonicalString()] =
+              model::DeepClone(field_filter.value());
+          return ValuesFrom(
+              std::move(field_value_map));  // NotIn/NotEqual is always a suffix
+        default:
+          continue;
+      }
+    }
+  }
+
+  return absl::nullopt;
+}
+
+absl::optional<IndexedBoundValues> Target::GetLowerBound(
+    const model::FieldIndex& field_index) {
+  std::vector<google_firestore_v1_Value> values;
+  bool inclusive = true;
+
+  // For each segment, retrieve a lower bound if there is a suitable filter or
+  // startAt.
+  for (const auto& segment : field_index.GetDirectionalSegments()) {
+    auto segment_bound = segment.kind() == Segment::Kind::kAscending
+                             ? GetAscendingBound(segment, start_at_)
+                             : GetDescendingBound(segment, start_at_);
+
+    if (!segment_bound.value.has_value()) {
+      // No lower bound exists
+      return absl::nullopt;
+    }
+
+    values.push_back(std::move(segment_bound.value.value()));
+    inclusive = (inclusive && segment_bound.is_inclusive);
+  }
+
+  return IndexedBoundValues{inclusive, std::move(values)};
+}
+
+absl::optional<IndexedBoundValues> Target::GetUpperBound(
+    const model::FieldIndex& field_index) {
+  std::vector<google_firestore_v1_Value> values;
+  bool inclusive = true;
+
+  // For each segment, retrieve an upper bound if there is a suitable filter or
+  // endAt.
+  for (const auto& segment : field_index.GetDirectionalSegments()) {
+    auto segment_bound = segment.kind() == Segment::Kind::kAscending
+                             ? GetDescendingBound(segment, end_at_)
+                             : GetAscendingBound(segment, end_at_);
+
+    if (!segment_bound.value.has_value()) {
+      // No lower bound exists
+      return absl::nullopt;
+    }
+
+    values.push_back(std::move(segment_bound.value.value()));
+    inclusive = (inclusive && segment_bound.is_inclusive);
+  }
+
+  return IndexedBoundValues{inclusive, std::move(values)};
+}
+
+IndexedBoundValue Target::GetAscendingBound(
+    const Segment& segment, const absl::optional<Bound>& bound) {
+  absl::optional<google_firestore_v1_Value> segment_value;
+  bool segment_inclusive = true;
+
+  // Process all filters to find a value for the current field segment
+  for (const auto& field_filter :
+       GetFieldFiltersForPath(segment.field_path())) {
+    absl::optional<google_firestore_v1_Value> filter_value;
+    bool filter_inclusive = true;
+
+    switch (field_filter.op()) {
+      case FieldFilter::Operator::LessThan:
+      case FieldFilter::Operator::LessThanOrEqual:
+        filter_value =
+            *model::GetLowerBound(field_filter.value().which_value_type);
```

Comment:
This could now theoretically cause a segfault if  GetLowerBound cleans up the memory location before the return type is used. My suggested fix would be to return just a nanopb Value in GetLowerBound and GetUpperBound and make sure that is does not contain any dynamically allocated memory (which we should be able to ensure).

Reply:
We can do that - but now we are back in the same situation where we always have to make a copy of the field value when we compute the bound, even if the bound is a value in the Target. Is that correct?